docs: Refactor feature-modules.md to make it a tutorial (#37948)
angular/angular: aio/content/guide/feature-modules.md

You can also declare [services](guide/glossary#service "Definition of service") that the NgModule can add to your app, including services from outside sources.

For more about NgModules, see [NgModules](guide/ngmodules "NgModules").

<div class="alert is-helpful">

For the final sample app with the new NgModule that this page describes, see the <live-example></live-example>.

</div>

## Objectives

* Create a new NgModule with its own component, which demonstrates how you can keep code related to a specific functionality or feature separate from other code.
* Import the new NgModule into your app to incorporate the specific functionality or feature with your app.
* Render the new NgModule component's template so that its markup appears within the app's main view.

## Prerequisites

To complete this tutorial, you should have a basic understanding of the following:

* [Typescript](guide/glossary#typescript "Definition of Typescript") and HTML5 programming
* [Templates](guide/glossary#template "Definition of a template") in HTML with CSS styles
* [Components](guide/glossary#component "Definition of a component")
* [Angular CLI](cli "Angular CLI")

> You might find the [Tour of Heroes tutorial](tutorial "Tour of Heroes") helpful, but it is not required.

Author: Fixed manually, appears in the next commit.

> ## Create an app project

Author: Using "Create a new app", thanks! Fixed manually, appears in the next commit.

To create an app just like the example for this tutorial, generate a new app workspace and project with the [Angular CLI](cli "Angular CLI").

<div class="alert is-helpful">

You can use an existing app workspace and project such as the one created in the Tour of Heroes (in [Create a new project](tutorial/toh-pt0 "Create a new project")), and skip to the next section, but it won't look exactly like the provided example.

> </div>

Author: Fixed manually, appears in the next commit.

Follow these steps:

1. Ensure that you are not already in an Angular workspace folder.
For example, if you have previously created the Getting Started workspace, change to the parent of that folder in order to create a new workspace.

2. Run the CLI command `ng new` and provide the name `feature-modules` as shown:

   <code-example language="sh" class="code-shell">
       ng new feature-modules
   </code-example>

   The command displays the request `Would you like to add Angular routing?`

>    Respond by typing **N** (no routing).

Author: Also fixed the list numbering so that these instructions are numbered underneath the list item.

   The command then displays a second request: `Which stylesheet format would you like to use?`

>    Press Enter (Return) to accept the default stylesheet.

Author: Fixed manually, appears in the next commit.

replace all of the content generated by the Angular CLI with the following HTML markup that uses Angular [interpolation](guide/glossary#interpolation "Definition of interpolation") for the component's `title` property:

   <code-example path="feature-modules/src/app/app.component.html" region="app-component-template-org" header="app.component.html (template)">
   </code-example>

5. Open the app module in `src/app/app.module.ts` and add the [`HttpClientModule`](api/common/http/HttpClientModule "HttpClientModule API") and [`FormsModule`](api/forms/FormsModule "FormsModule API") imports to the import section (which already includes `NgModule` and `BrowserModule`):

   <code-example path="feature-modules/src/app/app.module.ts" region="app-module-import"
   header="app.module.ts">
   </code-example>

   You import `FormsModule` so that Angular recognizes and applies the [`ngModel`](api/forms/NgModel "NgModel directive") directive.
   You import `HttpClientModule` to make available to your app Angular's [`HttpClient`](api/common/http/HttpClient "HttpClient class") mechanism for [communicating with a remote server over HTTP](guide/http "Communicating with backend services using HTTP").

>    Both are needed for the example app to work, but are not required for creating NgModules.

Author: Fixed manually, appears in the next commit.

> 6. In the same file, add `HttpClientModule` and `FormsModule` to the imports array in the `NgModule` metadata (which already includes `BrowserModule`):

Author: Used the following:

> In the same file, add `HttpClientModule` and `FormsModule` to the imports array in the `@NgModule()` metadata, which already includes `BrowserModule`:

```typescript

imports: [
    BrowserModule,
    FormsModule,
    HttpClientModule
  ],

```

## Generate the NgModule

Follow these steps:

> 1. Change to the workspace root folder (`feature-modules`).

Author: Fixed manually, appears in the next commit.

2. Enter the following CLI command, using a name such as `CustomerDashboard` as the name of your new NgModule (omit the `Module` suffix because the CLI automatically appends it to the name):

```sh
ng generate module CustomerDashboard

```

The command displays a request to share anonymous usage data about this project.
You can enter **Y** for yes, or press Return (Enter) or enter **N** for no.

The CLI creates the `customer-dashboard` folder with the `customer-dashboard.module.ts` file inside it containing the following:

```typescript

import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';

@NgModule({
  declarations: [],
  imports: [
    CommonModule
  ]
})
export class CustomerDashboardModule { }

```

The `customer-dashboard.module.ts` file includes the following import statements at the top:

* `NgModule`, which lets you use the `@NgModule` decorator to define an NgModule.
* `CommonModule`, which contributes directives such as `ngIf` and `ngFor` that are commonly used in templates.

> Note that new NgModules should import `CommonModule` rather than `BrowserModule`, which is  imported only once in the [root](guide/bootstrapping "Launching your app with a root module") NgModule and configures the entire app for the browser.

Author: I fixed this as above, but also made it accurate, as "you" don't import, the `ng new` command sets up that import. Here's the new version:

> Note that for new NgModules, you should import `CommonModule` rather than `BrowserModule`.
> The [root NgModule](guide/bootstrapping "Launching your app with a root module"), generated by the CLI `ng new` command, already imports `BrowserModule` in order to configure the entire app for the browser.
>

> You use the (now empty) `declarations` array in the `customer-dashboard.module.ts` file to declare the components, directives, and pipes that will belong exclusively to this particular NgModule.

Author: Fixed manually, appears in the next commit.

## Generate a new component in the NgModule

> To add a component to the new NgModule, enter the following CLI command, specifying the generated folder for the NgModule (`customer-dashboard`) and the component name (`CustomerDashboard`):

Reviewer: ```suggestion
To add a component to the new NgModule, enter the following CLI command, which specifes the folder for the NgModule, `customer-dashboard`, and the component name `CustomerDashboard`:
```

Author: Fixed manually, appears in the next commit.

```sh
ng generate component customer-dashboard/CustomerDashboard

```

> This command generates a folder for the new component within the `customer-dashboard` folder, and updates the `customer-dashboard.module.ts` file with the `CustomerDashboardComponent` info:

Reviewer: You might be able to combine this sentence with the next (line 147) as they are covering very similar ground.

Author: I combined them as follows:

> This command adds the `CustomerDashboardComponent` to the import list at the top of the `customer-dashboard.module.ts` file, and also adds it to the `declarations` array, which lets Angular know to associate this new component with this NgModule.
> The command also generates a folder for the new component within the `customer-dashboard` folder, and updates the `customer-dashboard.module.ts` file with the `CustomerDashboardComponent` info:
>

<code-example path="feature-modules/src/app/customer-dashboard/customer-dashboard.module.ts" region="customer-dashboard-component" header="src/app/customer-dashboard/customer-dashboard.module.ts"></code-example>

The command adds the `CustomerDashboardComponent` to the import list at the top of the `customer-dashboard.module.ts` file, and also adds it to the `declarations` array, which lets Angular know to associate this new component with this NgModule.

If you edit the `declarations` array manually, keep in mind the following:

> * Specify only [declarables](guide/glossary#declarable "Definition of a declarable") (component, directives, and pipes) in the `declarations` array.

Reviewer: ```suggestion
* Specify only [declarables](guide/glossary#declarable "Definition of a declarable")&mdash;components, directives, and pipes&mdash;in the `declarations` array.
```

Author: Fixed manually, appears in the next commit.

> * Declare these classes in an NgModule _only if_ they belong to that NgModule.

Reviewer: ```suggestion
* Declare these classes in an NgModule _only_ if they belong to that NgModule.
```

Did you mean to include the "if"?

Author: I did, but it doesn't matter, because either way is correct, so I chose your suggestion because it looks better.

* Do _not_ include other NgModules, Javascript modules, service classes, or non-Angular classes and objects such as strings, numbers, functions, entity models, configurations, business logic, and helper classes.
* Do _not_ add the same declarables to more than one NgModule.

> * Do _not_ declare a class that's already declared in another NgModule, whether it be in the root NgModule, another NgModule, or a third-party NgModule.

Reviewer: Bravo for the subjunctive! 🎉

For example, don't declare `FORMS_DIRECTIVES` from `@angular/forms` because the `FormsModule` already declares it.

Note that the `CustomerDashboardModule` export at the bottom of `customer-dashboard.module.ts` exposes the new NgModule so that other NgModules in your app can retrieve it, as shown in the next section.

## Incorporate the NgModule into your app

To incorporate the new NgModule into your app, you have to let the root NgModule, `app.module.ts`, know about it. 
Add `CustomerDashboardModule` to the imports in `app.module.ts` and to the `imports` array:

<code-example path="feature-modules/src/app/app.module.ts" region="app-module" header="src/app/app.module.ts"></code-example>

Now that the the root `AppModule` knows about the new NgModule, other NgModules in your app can use its implementation even though the new NgModule doesn't expose its components.

> If you add service providers to the new NgModule, `AppModule` would know about those too, so that they can be used by other NgModules.

Reviewer: You could link to the [hierarchical injectors guide](https://angular.io/guide/hierarchical-dependency-injection) here or the [DI intro page](.io/guide/dependency-injection). The `ModuleInjector` is the reason for this behavior.

Author: Adding context as well as the DI intro page link. The paragraph is now as follows:

> Now that the the root `AppModule` knows about the new NgModule, other NgModules in your app can use its implementation even though the new NgModule doesn't expose its components.
> You can add service providers to the new NgModule, as described in [Providing dependencies in modules](guide/providers "Providing dependencies in modules"). 
> `AppModule` would know about those too, so that they can be used by other NgModules.
> For more about service dependency injection, see [Dependency injection in Angular](guide/dependency-injection "Dependency injection in Angular").
>

## Render the new component's template

> When the CLI generated the `CustomerDashboardComponent` for the new NgModule, the component included a template, `customer-dashboard.component.html`, with the following HTML markup:

Reviewer: ```suggestion
When the CLI generated the `CustomerDashboardComponent` for the new NgModule, the component included a template, `customer-dashboard.component.html`, with the following HTML:
```

You can omit the word `markup` because it is the `M` in `HTML` (Like "ATM machine").

Author: Fixed manually, appears in the next commit.

<code-example path="feature-modules/src/app/customer-dashboard/customer-dashboard/customer-dashboard.component.html" region="feature-template" header="src/app/customer-dashboard/customer-dashboard/customer-dashboard.component.html"></code-example>

To display this markup as part of `AppComponent`, follow these steps:

1. Export the `CustomerDashboardComponent` in the `CustomerDashboardModule`.
   In `customer-dashboard.module.ts`, just beneath the `declarations` array, add an `exports` array containing `CustomerDashboardComponent`:

   <code-example path="feature-modules/src/app/customer-dashboard/customer-dashboard.module.ts" region="component-exports" header="src/app/customer-dashboard/customer-dashboard.module.ts"></code-example>

2. In the `AppComponent` template, `app.component.html`, add the  `<app-customer-dashboard>` tag:

   <code-example path="feature-modules/src/app/app.component.html" region="app-component-template" header="src/app/app.component.html"></code-example>

3. Run the app.
For example, use the CLI command `ng serve --open`.

> As a result, the `CustomerDashboardComponent` template renders along with the title (which renders by default):

Reviewer: ```suggestion
As a result, the `CustomerDashboardComponent` template renders along with the title:
```

Since we added this `title` above, we configured it to be that way. "Default" implies that the system creates and assumes this configuration. The way you've written it, without the parenthetical, is already clear.

Author: Fixed manually, appears in the next commit.

<div class="lightbox">
  <img src="generated/images/guide/feature-modules/feature-module.png" alt="feature module component">
</div>

<div class="alert is-helpful">

For the final sample app with the new NgModule that this page describes, see the <live-example></live-example>.

</div>

> ## Next steps

Reviewer: ```suggestion
## What's next
```

Author: Fixed manually, appears in the next commit.

This tutorial shows you how to create a new NgModule and incorporate it into your app.
You may also be interested in the following:

* For more about NgModules, see [NgModules](guide/ngmodules "NgModules").

> * To learn more about the root NgModule, see [Launching an app with a root module](guide/bootstrapping "Launching an app with a root module").

Reviewer: ```suggestion
* To learn more about the root NgModule, see [Launching your app with a root module](guide/bootstrapping "Launching an app with a root module").
```

Author: Fixed manually, appears in the next commit.
